File: CPP/src/scanner.cpp
#include "../include/scanner.hpp"
#include <cctype>
#include <optional>
#include <string>
#include <vector>

Scanner::Scanner(const std::string &src) : m_src(src) {}

std::vector<Token> Scanner::scan_tokens() & {
    while (!is_at_end()) {
        m_start = m_current;
        scan_token();
    }
    m_tokens.emplace_back(Token{EndOfFile, std::nullopt, ""});
    return m_tokens;
}

void Scanner::scan_token() & {
    char c = advance();
    switch (c) {
    case '(':
      add_null_token(LEFT_PAREN);
      break;
    case ')':
      add_null_token(RIGHT_PAREN);
      break;
    case '{':
      add_null_token(LEFT_BRACE);
      break;
    case '}':
        add_null_token(RIGHT_BRACE);
        break;
    case ',':
        add_null_token(COMMA);
        break;
    case '.':
        add_null_token(DOT);
        break;
    case '-':
        add_null_token(MINUS);
        break;
    case '+':
        add_null_token(PLUS);
        break;
    case ';': // COMMENT
        while (peek() != '\n' && peek() != '\0') {
            advance();
        }
        add_null_token(SEMICOLON);
        break;
    case '/':
        add_null_token(SLASH);
        break;
    case '*':
        add_null_token(STAR);
        break;

    case ' ':
    case '\t':
    case '\n':
    case '\r':
        break;
      
    case '\"':
        str();
        break;

    default:
      if (isdigit(c)) {
          number();
      } else if (isalpha(c)) {
          identifier();
      } else {
          std::cerr << "Unexpected character: " << c << '\n';
          return;
      }
    }
}

inline bool Scanner::is_at_end() const& noexcept { return m_current >= m_src.size(); }
inline char Scanner::advance() & { return m_src.at(m_current++); }

void Scanner::add_null_token(const Token_Type type) & noexcept {
    add_token(type, std::nullopt);
}

void Scanner::add_token(const Token_Type type, const lit &literal) & {
    const std::string str = m_src.substr(m_start, m_current - m_start);
    m_tokens.push_back({type, literal, str});
}

bool Scanner::match(const char expected) & {
    if (is_at_end()) {
        return false;
    } else if (m_src.at(m_current) != expected) {
        return false;
    } else {
        m_current++;
        return true;
    }
}

char Scanner::peek() const& {
    if (is_at_end()) {
        return '\0';
    } else {
        return m_src.at(m_current);
    }
}

char Scanner::peek_next() const& {
    if (m_current + 1 >= m_src.size()) {
        return '\0';
    } else {
        return m_src.at(m_current + 1);
    }
}

void Scanner::str() & {
    while (peek() != '\"' && !is_at_end()) {
        advance();
    }
    if (is_at_end()) {
        std::cerr << "Error: Unterminated string!\n";
        return;
    } else {
        advance();
        std::string str{m_src.substr(m_start + 1, m_current - m_start - 2)};
        add_token(STRING, str);
    }
}

void Scanner::number() & {
    while (isdigit(peek())) {
        advance();
    }
    if (peek() == '.' && isdigit(peek_next())) {
      advance();
    }
    while (isdigit(peek())) {
        advance();
    }
    std::string str_num{m_src.substr(m_start, m_current - m_start)};
    double number{std::stod(str_num)};
    add_token(NUMBER, number);
}

void Scanner::identifier() & {
    while (isalnum(peek())) {
        advance();
    }
    std::string text{m_src.substr(m_start, m_current - m_start)};
    auto it = m_keywords.find(text);
    if (it != m_keywords.end()) {
        add_null_token(it->second);
    } else {
        add_null_token(IDENTIFIER);
    }
}

File: Zig/src/scanner.zig
const std = @import("std");
const Token = @import("token.zig");

const Scanner_error = error {
    untermed_str,
    unknown_token,
};

pub const Scanner = struct {
    const Self = @This();
    start: usize = 0,
    current: usize = 0,

    source:[] u8,
    tokens: std.ArrayList(Token.Token),

    const keywords = std.StaticStringMap(Token.Token_Type).initComptime(.{
        .{"if", Token.Token_Type.IF},
        .{"def", Token.Token_Type.DEF},
        .{"let", Token.Token_Type.LET},
        .{"fn", Token.Token_Type.FN},
        .{"nil", Token.Token_Type.NIL},
        .{"do", Token.Token_Type.DO},
        .{"quote", Token.Token_Type.QUOTE},
        .{"eval", Token.Token_Type.EVAL},
        .{"const", Token.Token_Type.CONST},
        .{"first", Token.Token_Type.FIRST},
        .{"rest", Token.Token_Type.REST},
        .{"list", Token.Token_Type.LIST},
        .{"println", Token.Token_Type.PRINTLN},
        .{"not", Token.Token_Type.NOT},
    });

    pub fn init(alloc: std.mem.Allocator, src: []u8) Scanner {
        return .{
            .tokens = std.ArrayList(Token.Token).init(alloc),
            .source = src
        };
    }

    pub fn deinit(self: *Self) void { self.tokens.deinit(); }

    fn is_at_end(self: *Self) bool { return self.current >= self.source.len; }

    fn advance(self: *Self) u8 {
        self.current += 1;
        return self.source[self.current - 1];
    }

    fn add_null_token(self: *Self, t: Token.Token_Type) !void { try self.add_token(t, Token.Literal.None);}

    fn add_token(self: *Self, t: Token.Token_Type, lit: Token.Literal) !void {
        const text = self.source[self.start .. self.current];
        const token = Token.Token {
            .type = t,
            .lexeme = text,
            .literal = lit,
        };
        try self.tokens.append(token);
    }

    fn match(self: *Self, expected: u8)  bool {
        if(self.is_at_end()) {
            return false;
        } else if(self.source[self.current] != expected) {
            return false;
        }
        self.current += 1;
        return true;
    }

    fn match_next(self: *Self, expected: u8) bool {
        if(self.is_at_end()) {
            return false;
        } else if(self.peek_next() != expected) {
            return false;
        }
        self.current += 1;
        return true;
    }

    fn peek(self: *Self) ?u8 {
        if(self.is_at_end()) { return null; }
        return self.source[self.current];
    }

    fn peek_next(self: *Self) ?u8 {
        if(self.current + 1 >= self.source.len) { return null; }
        return self.source[self.current + 1];
    }

    fn string(self: *Self) !void {
        var c: u8 = self.peek().?;
        while(c != '\"' and !self.is_at_end()) {
            c = self.advance();
        }
        if(self.is_at_end()) {
            std.debug.print("Error! Unterminated string!\n", .{});
            return Scanner_error.untermed_str;
        }
        const value: []const u8 = self.source[self.start + 1.. self.current - 1];
        const literal = Token.Literal { .String = value };
        try self.add_token(Token.Token_Type.STRING, literal);
    }

    fn number(self: *Self) !void {
        while(std.ascii.isDigit(self.peek().?)) {
            _ = self.advance();
        }
        if(self.peek() == '.' and std.ascii.isDigit(self.peek_next().?)) {
            _ = self.advance();
        }

        const num_string = self.source[self.start .. self.current];
        const flt: f64 = try std.fmt.parseFloat(f64, num_string);
        const literal = Token.Literal{ .Float = flt };
        try self.add_token(Token.Token_Type.NUMBER, literal);
    }

    fn identifier(self: *Self) !void {
        while (self.peek() != null and (std.ascii.isAlphabetic(self.peek().?) or self.peek() == '_')) {
            _ = self.advance();
        }

        const text: []u8 = self.source[self.start .. self.current];
        const ttype_opt = keywords.get(text);
        const ttype = if (ttype_opt == null) Token.Token_Type.IDENTIFIER else ttype_opt.?;
        try self.add_null_token(ttype);
    }

    fn scan_token(self: *Self) !void {
        const c: u8 = self.advance();
        switch (c) {
            '(' => try self.add_null_token(Token.Token_Type.LEFT_PAREN),
            ')' => try self.add_null_token(Token.Token_Type.RIGHT_PAREN),
            ',' => try self.add_null_token(Token.Token_Type.COMMA),
            '.' => try self.add_null_token(Token.Token_Type.DOT),
            '-' => try self.add_null_token(Token.Token_Type.MINUS),
            '+' => try self.add_null_token(Token.Token_Type.PLUS),
            '*' => try self.add_null_token(Token.Token_Type.STAR),
            '/' => try self.add_null_token(Token.Token_Type.SLASH),

            '!' => if (self.match_next('=')) {
                try self.add_null_token(Token.Token_Type.BANG_EQUAL);
            } else {
                try self.add_null_token(Token.Token_Type.BANG);
            },

            '=' => if(self.match_next('=')) {
                try self.add_null_token(Token.Token_Type.EQUAL_EQUAL);
            } else {
                try self.add_null_token(Token.Token_Type.EQUAL);
            },

            '<' => if(self.match_next('=')) {
                try self.add_null_token(Token.Token_Type.LESS_EQUAL);
            } else {
                try self.add_null_token(Token.Token_Type.LESS);
            },

            '>' => if(self.match_next('=')) {
                try self.add_null_token(Token.Token_Type.GREATER_EQUAL);
            } else {
                try self.add_null_token(Token.Token_Type.GREATER);
            },

            '\"' => try self.string(),

            ' ' => {},
            '\r' => {},
            '\t' => {},
            '\n' => {},

            else => {
                if(std.ascii.isDigit(c)) {
                    try self.number();
                } else if(std.ascii.isAlphabetic(c)) {
                    try self.identifier();
                } else {
                    return Scanner_error.unknown_token;
                }
            }
        }
    }

    pub fn scan_tokens(self: *Self) ![]const Token.Token {
        while(!self.is_at_end()) {
            self.start = self.current;
            try self.scan_token();
        }
        const eof = Token.Token {
            .type = Token.Token_Type.EOF,
            .lexeme = "EOF",
            .literal = Token.Literal.None
        };

        try self.tokens.append(eof);
        return self.tokens.items;
    }
};